ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -223,6 +228,124 @@ def __post_init__(self):
+        ds_split: datasets.Dataset,
+        start_index: int,
+        end_index: int,
+    ):
+        original_chunk_size = end_index - start_index
+        if original_chunk_size < self.args.chunk_split_threshold:
+            total_chunks = 1
+            chunk_size = original_chunk_size
+            print(
+                f"Chunk range [{start_index}, {end_index}) is too small to split further. Processing and uploading as a single chunk."
+            )
+        else:
+            total_chunks = self.args.max_chunk_split
+            chunk_size = math.ceil(original_chunk_size / total_chunks)
+        failed_chunk_ranges = []
+        print(
+            f"Processing and uploading {total_chunks} chunks for range [{start_index}, {end_index}) with chunk size {chunk_size}"
+        )
+        for i in range(start_index, end_index, chunk_size):
+            chunk_start = i
+            chunk_end = min(i + chunk_size, start_index + original_chunk_size)
+
+            ds_chunk = ds_split.select(range(chunk_start, chunk_end))
+            ds_chunk_name = f"chunk-range-{chunk_start:09d}-{chunk_end:09d}.parquet"
+            ds_chunk_hub_path = os.path.join(
+                str(self.args.upload_subset), split_name, ds_chunk_name
+            )
+            ds_chunk_cache_path = os.path.join(
+                self.cache_dir,
+                self.args.dataset_name.replace("/", "__"),
+                str(self.args.upload_subset),
+                split_name,
+                ds_chunk_name,
+            )
+            try:
+                if os.path.exists(ds_chunk_cache_path):
+                    print(
+                        f"Skipping chunk {ds_chunk_name} as it has already been processed and uploaded."
+                    )
+                    ds_chunk_processed = datasets.Dataset.from_parquet(
+                        ds_chunk_cache_path
+                    )
+                else:
+                    print(f"Processing chunk {ds_chunk_name}")
+                    ds_chunk_processed = self._process(ds_chunk)
+                    self._upload(ds_chunk_processed, ds_chunk_hub_path, split_name)
+                    ds_chunk_processed.to_parquet(ds_chunk_cache_path)
+                self.total_samples_processed[split_name] = (
+                    self.total_samples_processed.get(split_name, 0)
+                    + len(ds_chunk_processed)
+                )
+
+            except Exception as e:
+                print(f"Failed to upload chunk {ds_chunk_name}: {e}. Retrying later.")
+                if total_chunks == 1:
+                    print(
+                        f"Finished processing and uploading 0/1 chunks for range [{start_index}, {end_index})"
+                    )
+                    self.chunks_not_uploaded.append((start_index, end_index))
+                    return []
+                failed_chunk_ranges.append((chunk_start, chunk_end))
+        successful_chunks = self.args.max_chunk_split - len(failed_chunk_ranges)
```

Comment:
I'm confused about the logic here.

1. I don't get what `self.chunks_not_uploaded` does. It only seems to be used for logging. So if you only have 1 chunk and you fail, then you're done and you won't retry?

2. IIRC the logic you said you wanted to implement was to recursively use smaller and smaller chunk sizes until you can do it in one go. I don't think I see that recursion here. The only chunking I see is via `chunk_split_threshold` which is static.

3. Having special logic for `total_chunks == 1` vs the other chunk sizes is also a bad smell, unless there's a special reason I didn't understand.

Reply:
So i guess its not strictely resursive, but the behavior is more or less recursive. The idea of _dynamic_chunk is, we split up the input chunk into n chunks. We try processing each one. if the ith chunk fails to upload, we add it to the failed chunks array. At the end we return the array The higher order function process_and_upload_split basically handles retrying the failed chunks. The base case for _dynamic_chunk doesnt split input chunk into n chunks if chunk_split_threshold is not met. This prevents us from chunking all the way down to 1 sample (if we don't want to).